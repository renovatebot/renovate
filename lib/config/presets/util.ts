import { logger } from '../../logger';
import { ensureTrailingSlash } from '../../util/url';
import type { Preset } from './types';

export const PRESET_DEP_NOT_FOUND = 'dep not found';
export const PRESET_NOT_FOUND = 'preset not found';

export type PresetFetcher = (
  repo: string,
  fileName: string,
  endpoint: string
) => Promise<Preset>;

export type FetchPresetConfig = {
  pkgName: string;
  filePreset: string;
  presetPath?: string;
  endpoint: string;
  fetch: PresetFetcher;
};

export async function fetchPreset({
  pkgName,
  filePreset,
  presetPath,
  endpoint,
  fetch,
}: FetchPresetConfig): Promise<Preset | undefined> {
  // eslint-disable-next-line no-param-reassign
  endpoint = ensureTrailingSlash(endpoint);
  const [fileName, presetName, subPresetName] = filePreset.split('/');
  const pathPrefix = presetPath ? `${presetPath}/` : '';
  const buildFilePath = (name: string): string => `${pathPrefix}${name}`;
  let jsonContent: any | undefined;
  if (fileName === 'default') {
    try {
      jsonContent = await fetch(
        pkgName,
        buildFilePath('default.json'),
        endpoint
      );
    } catch (err) {
      if (err.message !== PRESET_DEP_NOT_FOUND) {
        throw err;
      }
      logger.debug('default.json preset not found - trying renovate.json');
      jsonContent = await fetch(
        pkgName,
        buildFilePath('renovate.json'),
        endpoint
      );
    }
  } else {
    jsonContent = await fetch(
      pkgName,
      buildFilePath(`${fileName}.json`),
      endpoint
    );
  }

  if (!jsonContent) {
    throw new Error(PRESET_DEP_NOT_FOUND);
  }
  if (presetName) {
    const preset = jsonContent[presetName];
    if (!preset) {
      throw new Error(PRESET_NOT_FOUND);
    }
    if (subPresetName) {
      const subPreset = preset[subPresetName];
      if (!subPreset) {
        throw new Error(PRESET_NOT_FOUND);
      }
      return subPreset;
    }
    return preset;
  }
  return jsonContent;
}
