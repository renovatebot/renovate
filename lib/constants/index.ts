export * from './platforms';
export type * from './category';
