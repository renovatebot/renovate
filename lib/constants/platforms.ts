export type PlatformId =
  | 'azure'
  | 'bitbucket'
  | 'bitbucket-server'
  | 'gitea'
  | 'github'
  | 'gitlab';

export const GITHUB_API_USING_HOST_TYPES = [
  'github',
  'github-releases',
  'github-tags',
  'pod',
  'hermit',
  'github-changelog',
];

export const GITLAB_API_USING_HOST_TYPES = [
  'gitlab',
  'gitlab-releases',
  'gitlab-tags',
  'gitlab-packages',
  'gitlab-changelog',
];

export const BITBUCKET_API_USING_HOST_TYPES = ['bitbucket', 'bitbucket-tags'];
