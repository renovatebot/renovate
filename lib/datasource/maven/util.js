const got = require('../../util/got');

module.exports = {
  downloadHttpProtocol,
};

function isMavenCentral(pkgUrl) {
  return pkgUrl.host === 'central.maven.org';
}

function isTemporalError(err) {
  return (
    err.code === 'ECONNRESET' ||
    err.statusCode === 429 ||
    (err.statusCode > 500 && err.statusCode < 600)
  );
}

function isHostError(err) {
  return err.code === 'ETIMEDOUT';
}

function isNotFoundError(err) {
  return err.code === 'ENOTFOUND' || err.statusCode === 404;
}

function isPermissionsIssue(err) {
  return err.statusCode === 401 || err.statusCode === 403;
}

function isConnectionError(err) {
  return err.code === 'ECONNREFUSED';
}

async function downloadHttpProtocol(pkgUrl, hostType = 'maven') {
  let raw;
  try {
    raw = await got(pkgUrl, { hostType });
  } catch (err) {
    if (isNotFoundError(err)) {
      logger.debug(`Url not found ${pkgUrl}`);
    } else if (isHostError(err)) {
      // istanbul ignore next
      logger.warn({ pkgUrl }, `Cannot connect to ${hostType} host`);
    } else if (isPermissionsIssue(err)) {
      logger.warn(
        { pkgUrl },
        'Dependency lookup unauthorized. Please add authentication with a hostRule'
      );
    } else if (isTemporalError(err)) {
      logger.info({ err }, `Temporary error requesting ${pkgUrl}`);
      if (isMavenCentral(pkgUrl)) {
        throw new Error('registry-failure');
      }
    } else if (isConnectionError(err)) {
      // istanbul ignore next
      logger.info({ pkgUrl }, 'Connection refused to maven registry');
    } else {
      logger.warn({ err }, `Unknown error requesting ${pkgUrl}`);
    }
    return null;
  }
  return raw.body;
}
