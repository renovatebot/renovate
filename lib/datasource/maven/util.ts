import url from 'url';
import got from '../../util/got';
import { logger } from '../../logger';

function isMavenCentral(pkgUrl: url.URL | string): boolean {
  return (
    (typeof pkgUrl === 'string' ? pkgUrl : pkgUrl.host) === 'central.maven.org'
  );
}

function isTemporalError(err: { code: string; statusCode: number }): boolean {
  return (
    err.code === 'ECONNRESET' ||
    err.statusCode === 429 ||
    (err.statusCode >= 500 && err.statusCode < 600)
  );
}

function isHostError(err: { code: string }): boolean {
  return err.code === 'ETIMEDOUT';
}

function isNotFoundError(err: { code: string; statusCode: number }): boolean {
  return err.code === 'ENOTFOUND' || err.statusCode === 404;
}

function isPermissionsIssue(err: { statusCode: number }): boolean {
  return err.statusCode === 401 || err.statusCode === 403;
}

function isConnectionError(err: { code: string }): boolean {
  return err.code === 'ECONNREFUSED';
}

export async function downloadHttpProtocol(
  pkgUrl: url.URL | string,
  hostType = 'maven'
): Promise<string | null> {
  let raw: { body: string };
  try {
    raw = await got(pkgUrl, {
      hostType,
      hooks: {
        beforeRedirect: [
          (options: any): void => {
            if (
              options.search &&
              options.search.indexOf('X-Amz-Algorithm') !== -1
            ) {
              // maven repository is hosted on amazon, redirect url includes authentication.
              // eslint-disable-next-line no-param-reassign
              delete options.auth;
            }
          },
        ],
      },
    });
  } catch (err) {
    const failedUrl = pkgUrl.toString();
    if (isNotFoundError(err)) {
      logger.debug({ failedUrl }, `Url not found`);
    } else if (isHostError(err)) {
      // istanbul ignore next
      logger.warn({ failedUrl }, `Cannot connect to ${hostType} host`);
    } else if (isPermissionsIssue(err)) {
      logger.warn(
        { failedUrl },
        'Dependency lookup unauthorized. Please add authentication with a hostRule'
      );
    } else if (isTemporalError(err)) {
      logger.info({ failedUrl, err }, 'Temporary error');
      if (isMavenCentral(pkgUrl)) {
        throw new Error('registry-failure');
      }
    } else if (isConnectionError(err)) {
      // istanbul ignore next
      logger.info({ failedUrl }, 'Connection refused to maven registry');
    } else {
      logger.warn({ failedUrl, err }, 'Unknown error');
    }
    return null;
  }
  return raw.body;
}
