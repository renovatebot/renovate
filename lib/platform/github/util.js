module.exports = {
  expandError,
};

function expandError(err) {
  return {
    err,
    message: err.message,
    body: err.response ? err.response.body : undefined,
  };
}
