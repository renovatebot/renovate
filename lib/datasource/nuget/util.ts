import url from 'url';
import { logger } from '../../logger';
import { regEx } from '../../util/regex';

export function parseRegistryUrl(registryUrl: string): {
  feedUrl: string;
  protocolVersion: number;
} {
  try {
    const parsedUrl = url.parse(registryUrl);
    let protocolVersion = 2;
    const protocolVersionRegExp = regEx(/#protocolVersion=(2|3)/);
    const protocolVersionMatch = protocolVersionRegExp.exec(parsedUrl.hash);
    if (protocolVersionMatch) {
      parsedUrl.hash = '';
      protocolVersion = Number.parseInt(protocolVersionMatch[1], 10);
    } else if (parsedUrl.pathname.endsWith('.json')) {
      protocolVersion = 3;
    }
    return { feedUrl: url.format(parsedUrl), protocolVersion };
  } catch (err) {
    logger.debug({ err }, `nuget registry failure: can't parse ${registryUrl}`);
    return { feedUrl: registryUrl, protocolVersion: null };
  }
}
