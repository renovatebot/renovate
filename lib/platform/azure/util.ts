import {
  GitPullRequest,
  GitRepository,
  GitStatusContext,
  PullRequestAsyncStatus,
  PullRequestStatus,
} from 'azure-devops-node-api/interfaces/GitInterfaces';
import { logger } from '../../logger';
import { HostRule, PrState } from '../../types';
import { GitOptions } from '../../types/git';
import { add } from '../../util/sanitize';
import { AzurePr } from './types';

export function getNewBranchName(branchName?: string): string {
  if (branchName && !branchName.startsWith('refs/heads/')) {
    return `refs/heads/${branchName}`;
  }
  return branchName;
}

export function getGitStatusContextCombinedName(
  context: GitStatusContext
): string | undefined {
  if (!context) {
    return undefined;
  }
  const combinedName = `${context.genre ? `${context.genre}/` : ''}${
    context.name
  }`;
  logger.trace(`Got combined context name of ${combinedName}`);
  return combinedName;
}

export function getGitStatusContextFromCombinedName(
  context: string
): GitStatusContext | undefined {
  if (!context) {
    return undefined;
  }
  let name = context;
  let genre;
  const lastSlash = context.lastIndexOf('/');
  if (lastSlash > 0) {
    name = context.substr(lastSlash + 1);
    genre = context.substr(0, lastSlash);
  }
  return {
    genre,
    name,
  };
}

export function getBranchNameWithoutRefsheadsPrefix(
  branchPath: string
): string | undefined {
  if (!branchPath) {
    logger.error(`getBranchNameWithoutRefsheadsPrefix(${branchPath})`);
    return undefined;
  }
  if (!branchPath.startsWith('refs/heads/')) {
    logger.trace(
      `The refs/heads/ name should have started with 'refs/heads/' but it didn't. (${branchPath})`
    );
    return branchPath;
  }
  return branchPath.substring(11, branchPath.length);
}

export function getBranchNameWithoutRefsPrefix(
  branchPath?: string
): string | undefined {
  if (!branchPath) {
    logger.error(`getBranchNameWithoutRefsPrefix(${branchPath})`);
    return undefined;
  }
  if (!branchPath.startsWith('refs/')) {
    logger.trace(
      `The ref name should have started with 'refs/' but it didn't. (${branchPath})`
    );
    return branchPath;
  }
  return branchPath.substring(5, branchPath.length);
}

export function getRenovatePRFormat(azurePr: GitPullRequest): AzurePr {
  const number = azurePr.pullRequestId;
  const displayNumber = `Pull Request #${number}`;

  const sourceBranch = getBranchNameWithoutRefsheadsPrefix(
    azurePr.sourceRefName
  );
  const targetBranch = getBranchNameWithoutRefsheadsPrefix(
    azurePr.targetRefName
  );
  const body = azurePr.description;

  const createdAt = azurePr.creationDate?.toISOString();
  const state =
    {
      [PullRequestStatus.Abandoned]: PrState.Closed,
      [PullRequestStatus.Completed]: PrState.Merged,
    }[azurePr.status] || PrState.Open;

  const sourceRefName = azurePr.sourceRefName;

  const isConflicted = azurePr.mergeStatus === PullRequestAsyncStatus.Conflicts;

  return {
    ...azurePr,
    sourceBranch,
    state,
    number,
    displayNumber,
    body,
    sourceRefName,
    targetBranch,
    createdAt,
    ...(isConflicted && { isConflicted }),
  } as AzurePr;
}

export async function streamToString(
  stream: NodeJS.ReadableStream
): Promise<string> {
  const chunks: Uint8Array[] = [];
  /* eslint-disable promise/avoid-new */
  const p = await new Promise<string>((resolve, reject) => {
    stream.on('data', (chunk) => chunks.push(Buffer.from(chunk)));
    stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
    stream.on('error', (err) => reject(err));
  });
  return p;
}

function toBase64(from: string): string {
  return Buffer.from(from).toString('base64');
}

export function getStorageExtraCloneOpts(config: HostRule): GitOptions {
  let authType: string;
  let authValue: string;
  if (!config.token && config.username && config.password) {
    authType = 'basic';
    authValue = toBase64(`${config.username}:${config.password}`);
  } else if (config.token.length === 52) {
    authType = 'basic';
    authValue = toBase64(`:${config.token}`);
  } else {
    authType = 'bearer';
    authValue = config.token;
  }
  add(authValue);
  return {
    '-c': `http.extraheader=AUTHORIZATION: ${authType} ${authValue}`,
  };
}

export function max4000Chars(str: string): string {
  if (str && str.length >= 4000) {
    return str.substring(0, 3999);
  }
  return str;
}

export function getProjectAndRepo(
  str: string
): { project: string; repo: string } {
  logger.trace(`getProjectAndRepo(${str})`);
  const strSplit = str.split(`/`);
  if (strSplit.length === 1) {
    return {
      project: str,
      repo: str,
    };
  }
  if (strSplit.length === 2) {
    return {
      project: strSplit[0],
      repo: strSplit[1],
    };
  }
  const msg = `${str} can be only structured this way : 'repository' or 'projectName/repository'!`;
  logger.error(msg);
  throw new Error(msg);
}

export function getRepoByName(
  name: string,
  repos: GitRepository[]
): GitRepository | null {
  logger.trace(`getRepoByName(${name})`);

  let { project, repo } = getProjectAndRepo(name);
  project = project.toLowerCase();
  repo = repo.toLowerCase();

  return (
    repos?.find(
      (r) =>
        project === r?.project?.name?.toLowerCase() &&
        repo === r?.name?.toLowerCase()
    ) || null
  );
}
