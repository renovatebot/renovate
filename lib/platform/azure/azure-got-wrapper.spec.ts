import { PLATFORM_TYPE_AZURE } from '../../constants/platforms';
import * as _hostRules from '../../util/host-rules';

describe('platform/azure/azure-got-wrapper', () => {
  let azure: typeof import('./azure-got-wrapper');
  let hostRules: typeof _hostRules;
  beforeEach(() => {
    // reset module
    jest.resetModules();
    hostRules = require('../../util/host-rules');
    azure = require('./azure-got-wrapper');
  });

  describe('gitApi', () => {
    it('should throw an error if no config found', () => {
      expect(azure.gitApi).toThrow('No config found for azure');
      expect(azure.coreApi).toThrow('No config found for azure');
      expect(azure.policyApi).toThrow('No config found for azure');
    });
    it('should set personal access token and endpoint', () => {
      hostRules.add({
        hostType: PLATFORM_TYPE_AZURE,
        token: '1234567890123456789012345678901234567890123456789012',
        baseUrl: 'https://dev.azure.com/renovate1',
      });
      azure.setEndpoint('https://dev.azure.com/renovate1');

      const res = azure.azureObj();

      delete res.rest.client.userAgent;
      delete res.vsoClient.restClient.client.userAgent;

      // We will track if the lib azure-devops-node-api change
      expect(res).toMatchSnapshot();
    });
    it('should set bearer token and endpoint', () => {
      hostRules.add({
        hostType: PLATFORM_TYPE_AZURE,
        token: 'token',
        baseUrl: 'https://dev.azure.com/renovate2',
      });
      azure.setEndpoint('https://dev.azure.com/renovate2');

      const res = azure.azureObj();

      delete res.rest.client.userAgent;
      delete res.vsoClient.restClient.client.userAgent;

      // We will track if the lib azure-devops-node-api change
      expect(res).toMatchSnapshot();
    });

    it('should set password and endpoint', () => {
      hostRules.add({
        hostType: PLATFORM_TYPE_AZURE,
        username: 'user',
        password: 'pass',
        baseUrl: 'https://dev.azure.com/renovate3',
      });
      azure.setEndpoint('https://dev.azure.com/renovate3');

      const res = azure.azureObj();

      delete res.rest.client.userAgent;
      delete res.vsoClient.restClient.client.userAgent;

      // We will track if the lib azure-devops-node-api change
      expect(res).toMatchSnapshot();
    });
  });
});
