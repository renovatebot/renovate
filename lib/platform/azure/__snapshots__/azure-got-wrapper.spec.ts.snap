// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`platform/azure/azure-got-wrapper gitApi should set personal access token and endpoint 1`] = `
WebApi {
  "authHandler": PersonalAccessTokenCredentialHandler {
    "token": "1234567890123456789012345678901234567890123456789012",
  },
  "isNoProxyHost": [Function],
  "options": Object {
    "ignoreSslError": false,
  },
  "rest": RestClient {
    "client": HttpClient {
      "_allowRedirectDowngrade": false,
      "_allowRedirects": true,
      "_allowRetries": false,
      "_certConfig": undefined,
      "_disposed": false,
      "_httpProxy": undefined,
      "_ignoreSslError": false,
      "_keepAlive": false,
      "_maxRedirects": 50,
      "_maxRetries": 1,
      "_socketTimeout": undefined,
      "handlers": Array [
        PersonalAccessTokenCredentialHandler {
          "token": "1234567890123456789012345678901234567890123456789012",
        },
      ],
      "requestOptions": Object {
        "ignoreSslError": false,
      },
    },
  },
  "serverUrl": "https://dev.azure.com/renovate1",
  "vsoClient": VsoClient {
    "_initializationPromise": Promise {},
    "_locationsByAreaPromises": Object {},
    "basePath": "/renovate1",
    "baseUrl": "https://dev.azure.com/renovate1",
    "restClient": RestClient {
      "client": HttpClient {
        "_allowRedirectDowngrade": false,
        "_allowRedirects": true,
        "_allowRetries": false,
        "_certConfig": undefined,
        "_disposed": false,
        "_httpProxy": undefined,
        "_ignoreSslError": false,
        "_keepAlive": false,
        "_maxRedirects": 50,
        "_maxRetries": 1,
        "_socketTimeout": undefined,
        "handlers": Array [
          PersonalAccessTokenCredentialHandler {
            "token": "1234567890123456789012345678901234567890123456789012",
          },
        ],
        "requestOptions": Object {
          "ignoreSslError": false,
        },
      },
    },
  },
}
`;

exports[`platform/azure/azure-got-wrapper gitApi should set bearer token and endpoint 1`] = `
WebApi {
  "authHandler": BearerCredentialHandler {
    "token": "token",
  },
  "isNoProxyHost": [Function],
  "options": Object {
    "ignoreSslError": false,
  },
  "rest": RestClient {
    "client": HttpClient {
      "_allowRedirectDowngrade": false,
      "_allowRedirects": true,
      "_allowRetries": false,
      "_certConfig": undefined,
      "_disposed": false,
      "_httpProxy": undefined,
      "_ignoreSslError": false,
      "_keepAlive": false,
      "_maxRedirects": 50,
      "_maxRetries": 1,
      "_socketTimeout": undefined,
      "handlers": Array [
        BearerCredentialHandler {
          "token": "token",
        },
      ],
      "requestOptions": Object {
        "ignoreSslError": false,
      },
    },
  },
  "serverUrl": "https://dev.azure.com/renovate2",
  "vsoClient": VsoClient {
    "_initializationPromise": Promise {},
    "_locationsByAreaPromises": Object {},
    "basePath": "/renovate2",
    "baseUrl": "https://dev.azure.com/renovate2",
    "restClient": RestClient {
      "client": HttpClient {
        "_allowRedirectDowngrade": false,
        "_allowRedirects": true,
        "_allowRetries": false,
        "_certConfig": undefined,
        "_disposed": false,
        "_httpProxy": undefined,
        "_ignoreSslError": false,
        "_keepAlive": false,
        "_maxRedirects": 50,
        "_maxRetries": 1,
        "_socketTimeout": undefined,
        "handlers": Array [
          BearerCredentialHandler {
            "token": "token",
          },
        ],
        "requestOptions": Object {
          "ignoreSslError": false,
        },
      },
    },
  },
}
`;

exports[`platform/azure/azure-got-wrapper gitApi should set password and endpoint 1`] = `
WebApi {
  "authHandler": BasicCredentialHandler {
    "password": "pass",
    "username": "user",
  },
  "isNoProxyHost": [Function],
  "options": Object {
    "ignoreSslError": false,
  },
  "rest": RestClient {
    "client": HttpClient {
      "_allowRedirectDowngrade": false,
      "_allowRedirects": true,
      "_allowRetries": false,
      "_certConfig": undefined,
      "_disposed": false,
      "_httpProxy": undefined,
      "_ignoreSslError": false,
      "_keepAlive": false,
      "_maxRedirects": 50,
      "_maxRetries": 1,
      "_socketTimeout": undefined,
      "handlers": Array [
        BasicCredentialHandler {
          "password": "pass",
          "username": "user",
        },
      ],
      "requestOptions": Object {
        "ignoreSslError": false,
      },
    },
  },
  "serverUrl": "https://dev.azure.com/renovate3",
  "vsoClient": VsoClient {
    "_initializationPromise": Promise {},
    "_locationsByAreaPromises": Object {},
    "basePath": "/renovate3",
    "baseUrl": "https://dev.azure.com/renovate3",
    "restClient": RestClient {
      "client": HttpClient {
        "_allowRedirectDowngrade": false,
        "_allowRedirects": true,
        "_allowRetries": false,
        "_certConfig": undefined,
        "_disposed": false,
        "_httpProxy": undefined,
        "_ignoreSslError": false,
        "_keepAlive": false,
        "_maxRedirects": 50,
        "_maxRetries": 1,
        "_socketTimeout": undefined,
        "handlers": Array [
          BasicCredentialHandler {
            "password": "pass",
            "username": "user",
          },
        ],
        "requestOptions": Object {
          "ignoreSslError": false,
        },
      },
    },
  },
}
`;
